pkg/annotations/middleware.go

```diff
@@ -3,65 +3,102 @@ package annotations
 import (
 	"errors"
 	"fmt"
-	"strconv"
 
 	"github.com/containous/traefik/v2/pkg/config/dynamic"
 )
 
-const (
-	annotationRetryAttempts            = baseAnnotation + "retry-attempts"
-	annotationCircuitBreakerExpression = baseAnnotation + "circuit-breaker-expression"
-	annotationRateLimitAverage         = baseAnnotation + "ratelimit-average"
-	annotationRateLimitBurst           = baseAnnotation + "ratelimit-burst"
-)
+type middlewareBuilder func(middleware *dynamic.Middleware, annotations map[string]string) (bool, error)
 
 // BuildMiddleware builds a middleware from the given annotations
 func BuildMiddleware(annotations map[string]string) (*dynamic.Middleware, error) {
-	var middleware dynamic.Middleware
-
-	// Build circuit-breaker middleware.
-	if circuitBreakerExpression, ok := annotations[annotationCircuitBreakerExpression]; ok {
-		middleware.CircuitBreaker = &dynamic.CircuitBreaker{Expression: circuitBreakerExpression}
+	builders := []middlewareBuilder{
+		buildRetryMiddleware,
+		buildRateLimitMiddleware,
+		buildCircuitBreakerMiddleware,
 	}
 
-	// Build retry middleware.
-	if retryAttempts, ok := annotations[annotationRetryAttempts]; ok {
-		attempts, err := strconv.Atoi(retryAttempts)
+	var (
+		middleware        dynamic.Middleware
+		middlewareCounter int
+	)
+
+	for _, builder := range builders {
+		ok, err := builder(&middleware, annotations)
 		if err != nil {
-			return nil, fmt.Errorf("unable to build retry middleware, %q annotation is invalid: %w", annotationRetryAttempts, err)
+			return nil, err
 		}
 
-		middleware.Retry = &dynamic.Retry{Attempts: attempts}
+		if ok {
+			middlewareCounter++
+		}
 	}
 
-	// Build rate-limit middleware.
-	rateLimitAverage, hasRateLimitAverage := annotations[annotationRateLimitAverage]
-	rateLimitBurst, hasRateLimitBurst := annotations[annotationRateLimitBurst]
+	if middlewareCounter == 0 {
+		return nil, nil
+	}
 
-	if hasRateLimitAverage && hasRateLimitBurst {
-		average, err := strconv.Atoi(rateLimitAverage)
-		if err != nil {
-			return nil, fmt.Errorf("unable to build rate-limit middleware, %q annotation is invalid: %w", annotationRateLimitAverage, err)
-		}
+	return &middleware, nil
+}
 
-		burst, err := strconv.Atoi(rateLimitBurst)
-		if err != nil {
-			return nil, fmt.Errorf("unable to build rate-limit middleware, %q annotation is invalid: %w", annotationRateLimitBurst, err)
+func buildRetryMiddleware(middleware *dynamic.Middleware, annotations map[string]string) (bool, error) {
+	retryAttempts, errRetryAttempts := GetRetryAttempts(annotations)
```

Comment:
Why not call this `err`? 🤔 🤔 🤔 🤔

Reply:
It was useful when all these middleware were built in the same function. I guess I just forgot the change them back.